### PR TITLE
Merge branch 'master' of git://github.com/BowenFu/racket-cas into Bow…

### PR DESCRIPTION
**Pull request by BowenFu**
…enFu-master

### CODE REVIEW

File: racket-cas/racket-cas.rkt [hunk 1]
```diff
@@ -741,103 +758,118 @@
...
+;;;
+
+; For a rational number a/b the numerator and denominator are a and b respectively. 
+; For an expression u which happen be a ratio (a product of a rational and factors
+; of the form (expt u p) we compute the numerator and denominator.
+; For other expressions 1 is returned as a denominator.
+
+; When matching we will in some situations want to match a "real" quotient,
+; i.e. one where the denominator isn't 1. So we will define two different match
+; expanders ⊘ and Quotient. Here ⊘  matches only if the denominator isn't 1,
```
Comment (BowenFu) [inline racket-cas/racket-cas.rkt:773]: In my PR, (Quotient u v)will match a Sum (a/b + c/d). and the v is cd.

Looks like  now the denominator is 1.

File: racket-cas/racket-cas.rkt [hunk 2]
```diff
@@ -741,103 +758,118 @@
...
-  (check-equal? (math-match (⊘ x (⊗ 2 y z)) [(Quotient u v) (list u v)]) '(x (* 2 y z)))
-  (check-equal? (math-match (⊘ x y) [(Quotient u v) (list u v)]) '(x y))
-  (check-equal? (math-match (⊘ (⊗ y 3) x) [(Quotient u v) (list u v)]) '((* 3 y) x))
+  (displayln "TEST - QUOTIENT")
+  (check-equal? (math-match 2/3             [(Quotient u v) (list u v)] [_ #f]) '(2 3))
+  (check-equal? (math-match '(* 2.3 x)      [(Quotient u v) (list u v)] [_ #f]) '((* 2.3 x) 1))
+  (check-equal? (math-match (⊘ x (⊗ 2 y z)) [(Quotient u v) (list u v)] [_ #f]) '(x (* 2 y z)))
+  (check-equal? (math-match (⊘ x y)         [(Quotient u v) (list u v)] [_ #f]) '(x y))
+  (check-equal? (math-match (⊘ (⊗ y 3) x)   [(Quotient u v) (list u v)] [_ #f]) '((* 3 y) x))
+  ; (check-equal? (math-match (⊕ x 2/3)       [(Quotient u v) (list u v)] [_ #f]) '((+ 2 (* 3 x)) 3)) ; not simple
```
Comment (BowenFu) [inline racket-cas/racket-cas.rkt:823]: This fails because together is not conducted.

Comment (BowenFu) [inline racket-cas/racket-cas.rkt:823]: This fails because together is not conducted.

File: racket-cas/racket-cas.rkt [hunk 3]
```diff
@@ -848,174 +880,185 @@
...
+      [r          (- r)]
+      [(⊗ r u) (⊗ (- r) u)]
+      [_ (error)]))
+  (match u
+    [r #:when (and (real? r) (negative? r)) (values 0 (- r))]
+    [r #:when      (real? r)                (values r 0)]
+    [x #:when (symbol? x)                   (values x 0)] ; assume variables are positive
+    [(list* '+ us) (let loop ([us us] [pos '()] [neg '()])
+                     (match us
+                       ['()         (values (reverse-terms->sum pos)
```
Comment (BowenFu) [inline racket-cas/racket-cas.rkt:954]: Why reverse pos? It is already in normalized order?

File: racket-cas/racket-cas.rkt [hunk 4]
```diff
@@ -848,174 +880,185 @@
...
 
 ; used for quotient.
 ; first call together, then split up the term-numerator/denominator parts.
 ; differences can be indicated by the following tests.
-(define (numerator/denominator s)
-  (term-numerator/denominator (together s))
-  )
+(define (mma-numerator/denominator s)
+  (term-numerator/denominator (together s)))
+
```
Comment (BowenFu) [inline racket-cas/racket-cas.rkt:993]: quotient can use mma-numerator/denominator

File: racket-cas/racket-cas.rkt (hunk 1 above)
Comment (BowenFu) [reply racket-cas/racket-cas.rkt:773]: Hi @soegaard 
I have reviewed your changes, the main concern is about the Quotient.
No together conducted for it, is it what we need?

Comment (soegaard) [reply racket-cas/racket-cas.rkt:773]: I think it is a good idea to make the pattern matchers as fast as possible, so
users don't accidently use an expensive operation.
In the case of Quotient, a user can call `together` explicitly if he needs to.

But we could add, say, QuotientTogether and have both options?

File: racket-cas/racket-cas.rkt (hunk 2 above)
Comment (soegaard) [reply racket-cas/racket-cas.rkt:823]: Yes.

File: racket-cas/racket-cas.rkt (hunk 3 above)
Comment (soegaard) [reply racket-cas/racket-cas.rkt:954]: The loop accumulates the terms in pos and neg in reverse order.

File: racket-cas/racket-cas.rkt (hunk 4 above)
Comment (soegaard) [reply racket-cas/racket-cas.rkt:993]: I'll add an QuotientTogether so we have both options.

**Pull request closed without merging**